File: jsrc/legacy/SxxMachine/builtin/PRED_$end_exception_1.java
package SxxMachine.builtin;

import SxxMachine.Operation;
import SxxMachine.Predicate;
import SxxMachine.Prolog;
import SxxMachine.Term;

/**
 * <code>'$end_exception'/1<code><br>
 * @author Mutsunori Banbara (banbara@kobe-u.ac.jp)
 * @author Naoyuki Tamura (tamura@kobe-u.ac.jp)
 * @version 1.0
 */
class PRED_$end_exception_1 extends Predicate.P1 {
	public PRED_$end_exception_1(Term a1, Operation cont) {
		arg1 = a1;
		this.cont = cont;
	}

	public Operation exec(final Prolog engine) {
		engine.setB0();
		final int B = engine.popCatcherB();
		engine.trail.push(()->engine.pushCatcherB(B));
		return cont;
	}
}

File: src/legacy/SxxMachine/builtin/PRED_$end_exception_1.cpp
using namespace std;

#include "PRED_$end_exception_1.h"
#include "../../../machine/SxxMachine/Operation.h"
#include "../../../machine/SxxMachine/Term.h"
#include "../../../machine/SxxMachine/Prolog.h"

namespace SxxMachine
{
	using Operation = SxxMachine::Operation;
	using Predicate = SxxMachine::Predicate;
	using Prolog = SxxMachine::Prolog;
	using Term = SxxMachine::Term;

	PRED_$end_exception_1::PRED_$end_exception_1(Term *a1, Operation cont)
	{
		LARG[0] = a1;
		this->cont = cont;
	}

	Operation PRED_$end_exception_1::exec(Prolog *const engine)
	{
		engine->setB0();
		constexpr int B = engine->popCatcherB();
		engine->trail->push([&] ()
		{
		engine->pushCatcherB(B);
		});
		return cont;
	}
}
